ctapipe/instrument/camera.py

```diff
@@ -518,6 +533,69 @@ def get_shower_coordinates(self, x, y, psi):
 
         return longi, trans
 
+    def get_pixel_id(self, x, y):
+        '''
+        Return the camera pixel number which contains a given position (x,y) 
+        in the camera frame. The (x,y) coordinates can be arrays (of equal length),
+        for which the methods returns an array of pixel ids. A warning is raised if the 
+        position falls outside the camera.
+
+        Parameters
+        ----------
+        x: astropy.units.Quantity (distance) of horizontal position(s) in the camera frame
+        y: astropy.units.Quantity (distance) of vertical position(s) in the camera frame
+
+        Returns
+        -------
+        pixel_ids: Pixel number or array of pixel numbers. Returns -1 if position falls 
+                   outside camera
+        '''
+
+        if not np.all(self.pix_area == self.pix_area[0], axis=0):
+            logger.warning(" Method not implemented for cameras with varying pixel sizes")
+
+        points_searched = np.dstack([x.to_value(u.m), y.to_value(u.m)])
+
+        kdtree = self.all_pixels_kdtree
+        dist, pixel_ids = kdtree.query(points_searched)
```

Comment:
`distance_upper_bound` implemented in a34e9bd. I left the precise check for now (only done for points between circumference and actual surface), but the corresponding code can simply be removed (is only called if points haven't been sorted out already by `distance_upper_bound`). For many queried points, `distance_upper_bound` seems a factor 3 faster than the current exact check.

> At a certain size, it also might be faster to build a kdtree for the points that are looked for.

Absolutely, but I fear this might turn into some over-engineering?